File: machine/ts/render/video_presenter.ts
import { BFont } from './shared/bitmap_font';
import type { vec2 } from '../common/vector';
import type { BackendContext, GPUBackend, PresentationMode, RenderContext, TextureHandle } from './backend/backend';
import { RGBA8_LINEAR_TEXTURE_PARAMS, RGBA8_SRGB_TEXTURE_PARAMS } from './backend/texture_params';
import { RenderPassLibrary } from './backend/pass/library';
import { DeviceQuantizeMode } from './post/device_quantize/mode';
import { RenderGraphRuntime, type FrameData } from './graph/graph';
import type { VideoOutput, VideoSurface } from '../platform';
import type { GxGpuDeviceOutput } from '../machine/devices/gx/device_output';
import { renderGate } from '../common/taskgate';

export class VideoPresenter implements RenderContext {
	public readonly surface: VideoSurface;
	public accessor default_font: BFont;
	public viewportSize: vec2;
	public viewportScale = 1;
	public canvasSize: vec2;
	public canvasScale = 1;
	public readonly nativeCtx: BackendContext;
	public readonly backend: GPUBackend;
	public offscreenCanvasSize: vec2;
	public textures: { [k: string]: TextureHandle } = {};

	public enable_noise = true;
	public enable_colorbleed = true;
	public enable_scanlines = true;
	public enable_blur = true;
	public enable_glow = true;
	public enable_fringing = true;
	public enable_aperture = false;
	public show_resource_usage_gizmo = false;
	public noiseIntensity = 0.3;
	public colorBleed: [number, number, number] = [0.02, 0.0, 0.0];
	public blurIntensity = 0.6;
	public glowColor: [number, number, number] = [0.12, 0.10, 0.09];
	public crt_postprocessing_enabled = true;

	public presentationMode: PresentationMode = 'completed';
	public commitPresentationFrame = false;
	public presentationHistorySourceIndex: 0 | 1 = 0;

	private renderGraph!: RenderGraphRuntime;
	private pipelineRegistry!: RenderPassLibrary;
	private readonly frame: FrameData = { frameIndex: 0, time: 0, delta: 0 };
	private _deviceQuantizeMode = DeviceQuantizeMode.None;
	private _deviceQuantizeConfigurationRevision = 0;

	constructor(private readonly output: VideoOutput, backend: GPUBackend, viewportWidth: number, viewportHeight: number) {
		this.surface = output.surface;
		this.backend = backend;
		this.nativeCtx = backend.context;
		this.viewportSize = { x: viewportWidth, y: viewportHeight };
		this.canvasSize = { x: viewportWidth, y: viewportHeight };
		this.offscreenCanvasSize = { x: viewportWidth, y: viewportHeight };
	}

	public get deviceQuantizeMode(): DeviceQuantizeMode {
		return this._deviceQuantizeMode;
	}

	public set deviceQuantizeMode(mode: DeviceQuantizeMode) {
		if (this._deviceQuantizeMode === mode) return;
		this._deviceQuantizeMode = mode;
		this._deviceQuantizeConfigurationRevision += 1;
	}

	public get deviceQuantizeConfigurationRevision(): number {
		return this._deviceQuantizeConfigurationRevision;
	}

	public get presentationHistoryDestinationIndex(): 0 | 1 {
		return this.presentationHistorySourceIndex === 0 ? 1 : 0;
	}

	public initialize(pipelineRegistry: RenderPassLibrary): void {
		this.pipelineRegistry = pipelineRegistry;
		this.surface.setRenderTargetSize(this.canvasSize.x, this.canvasSize.y);
		this.resetPresentationHistory();
		this._deviceQuantizeConfigurationRevision += 1;
		this.renderGraph = this.pipelineRegistry.buildRenderGraph();
	}

	public dispose(): void {
		this.renderGraph.dispose();
		this.pipelineRegistry.dispose();
		this.clearTextures();
	}

	public setRenderTargetSize(width: number, height: number): void {
		if (this.viewportSize.x === width && this.viewportSize.y === height) {
			return;
		}
		this.viewportSize.x = width;
		this.viewportSize.y = height;
		this.canvasSize.x = width;
		this.canvasSize.y = height;
		this.offscreenCanvasSize.x = width;
		this.offscreenCanvasSize.y = height;
		this.surface.setRenderTargetSize(width, height);
		const dimensions = this.output.getSize(this.viewportSize, this.canvasSize);
		this.viewportScale = dimensions.viewportScale;
		this.canvasScale = dimensions.canvasScale;
		this.rebuildGraph();
	}

	public present(output: GxGpuDeviceOutput, timeSeconds: number, deltaSeconds: number): void {
		if (!renderGate.ready) return;
		this.backend.beginFrame();
		try {
			this.frame.time = timeSeconds;
			this.frame.delta = deltaSeconds;
			this.renderGraph.execute(this.frame, output);
			this.frame.frameIndex = (this.frame.frameIndex + 1) >>> 0;
			this.finalizePresentation();
		} finally {
			this.backend.endFrame();
		}
	}

	public clearTextures(): void {
		for (const name in this.textures) {
			this.backend.destroyTexture(this.textures[name]);
		}
		this.textures = {};
	}

	public initializeDefaultTextures(): void {
		this.clearTextures();
		this.textures['_default_albedo'] = this.backend.createSolidTexture2D(1, 1, 0xffffffff, RGBA8_SRGB_TEXTURE_PARAMS);
		this.textures['_default_normal'] = this.backend.createSolidTexture2D(1, 1, 0xff7f7fff, RGBA8_LINEAR_TEXTURE_PARAMS);
		this.textures['_default_mr'] = this.backend.createSolidTexture2D(1, 1, 0xffffffff, RGBA8_LINEAR_TEXTURE_PARAMS);
	}

	public configurePresentation(mode: PresentationMode, commitFrame: boolean): void {
		this.presentationMode = mode;
		this.commitPresentationFrame = commitFrame;
	}

	public rebuildGraph(): void {
		this.renderGraph.dispose();
		this.resetPresentationHistory();
		this._deviceQuantizeConfigurationRevision += 1;
		this.renderGraph = this.pipelineRegistry.buildRenderGraph();
	}

	private resetPresentationHistory(): void {
		this.presentationMode = 'completed';
		this.commitPresentationFrame = false;
		this.presentationHistorySourceIndex = 0;
	}

	private finalizePresentation(): void {
		if (!this.commitPresentationFrame) {
			return;
		}
		this.presentationHistorySourceIndex = this.presentationHistoryDestinationIndex;
	}
}

File: machine/cpp/render/video_presenter.cpp
/*
 * video_presenter.cpp - Host video presentation for BMSX
 *
 * Routes host/editor render submissions to render queues.
 */

#include "video_presenter.h"
#include "backend/pass/library.h"
#include "graph/graph.h"
#include <stdexcept>
#include <utility>

namespace bmsx {

/* ============================================================================
 * VideoPresenter implementation
 * ============================================================================ */

VideoPresenter::VideoPresenter(VideoOutput& output, std::unique_ptr<GPUBackend> backend, i32 viewportWidth, i32 viewportHeight)
	: viewportSize{static_cast<f32>(viewportWidth), static_cast<f32>(viewportHeight)}
	, canvasSize{static_cast<f32>(viewportWidth), static_cast<f32>(viewportHeight)}
	, offscreenCanvasSize{static_cast<f32>(viewportWidth), static_cast<f32>(viewportHeight)}
	, m_output(output)
	, m_backend(std::move(backend))
{
}

VideoPresenter::~VideoPresenter() {
	m_renderGraph.reset();
	m_pipelineRegistry.reset();
	clearTextures();
}

void VideoPresenter::setScanoutSize(i32 width, i32 height) {
	if (!fixedRenderTargetSize) setRenderTargetSize(width, height);
	if (offscreenCanvasSize.x == static_cast<f32>(width) && offscreenCanvasSize.y == static_cast<f32>(height)) return;
	offscreenCanvasSize = {static_cast<f32>(width), static_cast<f32>(height)};
	rebuildGraph();
}

void VideoPresenter::setFixedRenderTargetSize(i32 width, i32 height) {
	fixedRenderTargetSize = true;
	setRenderTargetSize(width, height);
}

void VideoPresenter::useScanoutRenderTargetSize() {
	fixedRenderTargetSize = false;
	setRenderTargetSize(static_cast<i32>(offscreenCanvasSize.x), static_cast<i32>(offscreenCanvasSize.y));
}

void VideoPresenter::setRenderTargetSize(i32 width, i32 height) {
	if (viewportSize.x == static_cast<f32>(width) && viewportSize.y == static_cast<f32>(height)) {
		return;
	}
	viewportSize = {static_cast<f32>(width), static_cast<f32>(height)};
	canvasSize = viewportSize;
	m_backend->resizePresentationTarget(width, height);
	m_output.setDisplaySize(width, height);
}
void VideoPresenter::clearTextures() {
	for (const auto& entry : textures) {
		m_backend->destroyTexture(entry.second);
	}
	textures.clear();
}

void VideoPresenter::initializeDefaultTextures() {
	clearTextures();
	textures["_default_albedo"] = m_backend->createSolidTexture2D(1, 1, 0xffffffffu, RGBA8_SRGB_TEXTURE_PARAMS);
	textures["_default_normal"] = m_backend->createSolidTexture2D(1, 1, 0xff7f7fffu, RGBA8_LINEAR_TEXTURE_PARAMS);
	textures["_default_mr"] = m_backend->createSolidTexture2D(1, 1, 0xffffffffu, RGBA8_LINEAR_TEXTURE_PARAMS);
}

void VideoPresenter::configurePresentation(PresentationMode mode, bool commitFrame) {
	presentationMode = mode;
	commitPresentationFrame = commitFrame;
}

void VideoPresenter::setDeviceQuantizeMode(DeviceQuantizeMode mode) {
	if (m_deviceQuantizeMode == mode) {
		return;
	}
	m_deviceQuantizeMode = mode;
	m_deviceQuantizeConfigurationRevision += 1u;
}

void VideoPresenter::resetPresentationHistory() {
	m_committedSequence.reset();
	presentationMode = PresentationMode::Completed;
	commitPresentationFrame = false;
	presentationHistorySourceIndex = 0;
}

void VideoPresenter::finalizePresentation() {
	if (!commitPresentationFrame) {
		return;
	}
	presentationHistorySourceIndex = presentationHistoryDestinationIndex();
	m_committedSequence = m_renderFrameIndex;
}

RgbaImage VideoPresenter::captureGameFrame() {
	if (!m_committedSequence) throw std::runtime_error("No completed game frame is available.");
	return m_renderGraph->captureHistoryTexture(presentationHistorySourceIndex);
}

/**
 * Main render loop - executes the render graph.
 *
 * The render graph executes the GX GPU, post, and host passes
 * in the correct order.
 */
void VideoPresenter::present(const GxGpuDeviceOutput& output, f64 timeSeconds, f64 deltaSeconds) {
	m_backend->beginFrame();

	FrameData frame;
	frame.frameIndex = m_renderFrameIndex;
	frame.time = timeSeconds;
	frame.delta = deltaSeconds;
	m_renderGraph->execute(&frame, output);
	m_renderFrameIndex += 1u;
	finalizePresentation();
	m_backend->endFrame();
}

void VideoPresenter::installRenderPipeline(std::unique_ptr<RenderPassLibrary> registry) {
	m_renderGraph.reset();
	m_pipelineRegistry = std::move(registry);
	resetPresentationHistory();
	m_deviceQuantizeConfigurationRevision += 1u;
	m_renderGraph = m_pipelineRegistry->buildRenderGraph();
}

void VideoPresenter::releaseRenderPipeline() {
	m_renderGraph.reset();
	m_pipelineRegistry.reset();
}

// ─────────────────────────────────────────────────────────────────────────────
// Render graph
// ─────────────────────────────────────────────────────────────────────────────

void VideoPresenter::rebuildGraph() {
	m_renderGraph.reset();
	resetPresentationHistory();
	m_deviceQuantizeConfigurationRevision += 1u;
	m_renderGraph = m_pipelineRegistry->buildRenderGraph();
}

} // namespace bmsx
